nip45: add hyperloglog relay response

fiatjaf opened the pull request:
Here's a nice colorful video explanation of HyperLogLog: https://www.youtube.com/watch?v=lJYufx0bfpw
And here's a very interesting article with explanations, graphs and other stuff: http://antirez.com/news/75

If relays implement this we can finally get follower counts that do not suck and without having to use a single relay (aka relay.nostr.band) as the global source of truth for the entire network -- at the same time as we save the world by consuming an incomparably small fraction of the bandwidth.

Even if one was to download 2 reaction events in order to display a silly reaction count number in a UI that would already be using more bytes than this HLL value does (actually considering deflate compression the COUNT response with the HLL value is already smaller than a single reaction EVENT response).

This requires trusting relays to not lie about the counts and the HLL values, but this NIP always required that anyway, so no change there.

---

HyperLogLog can be implement in multiple ways, with different parameters and whatnot. Luckily most of the customizations (for example, the differences between HyperLogLog++ and HyperLogLog) can be applied at the final step, so it is a client choice. This NIP only describes the part that is needed for interoperability, which is how relays should compute the values and then return them to clients.

Because implementations would have to agree on parameters such as the number of registers to use, this NIP also fixes that number in 256 for simplicity's sake (makes it simpler implement since it's the maximum value of one byte) and also because it is a reasonable amount.

---

These are some random estimations I did, to showcase how efficient those 256 bytes can be:

```
real count: 2 estimation: 2
real count: 4 estimation: 4
real count: 6 estimation: 6
real count: 7 estimation: 7
real count: 12 estimation: 12
real count: 15 estimation: 15
real count: 22 estimation: 20
real count: 36 estimation: 36
real count: 44 estimation: 43
real count: 47 estimation: 44
real count: 64 estimation: 65
real count: 77 estimation: 72
real count: 89 estimation: 88
real count: 95 estimation: 93
real count: 104 estimation: 101
real count: 116 estimation: 113
real count: 122 estimation: 131
real count: 144 estimation: 145
real count: 150 estimation: 154
real count: 199 estimation: 196
real count: 300 estimation: 282
real count: 350 estimation: 371
real count: 400 estimation: 428
real count: 500 estimation: 468
real count: 600 estimation: 595
real count: 777 estimation: 848
real count: 922 estimation: 922
real count: 1000 estimation: 978
real count: 1500 estimation: 1599
real count: 2222 estimation: 2361
real count: 9999 estimation: 10650
real count: 13600 estimation: 13528
real count: 80000 estimation: 73439
real count: 133333 estimation: 135973
real count: 200000 estimation: 189470
```

As you can see they are almost perfect for small counts, but still pretty good for giant counts.

fiatjaf commented:
Implementation here: https://github.com/nbd-wtf/go-nostr/blob/4532e79d8e52d105888b41d2563c21c8104ea510/nip45/hll.go

alexgleason commented:
Well, that's pretty mind-blowing. :exploding_head: 

:thinking: 

It weirdly incentives PoW on the middle of the ID. The relay is responsible for preventing that. Existing measures (WoT) might be enough for some relays. But the default behavior is to trust not just the relay operator but also the users of that relay to not PoW the middle of the ID.

alexgleason commented:
Why shouldn't the relay just generate a random number every time instead of inspecting the event itself? Then it's not idempotent but it prevents abuse.

fiatjaf commented:
You can't generate a random number because each distinct item must yield the same number, therefore you need a hash of the item. The event id is already that hash, but yes, people can do PoW to make their reaction count more or something like that.

I considered letting each relay pick a different random offset of the id to count from, which would mitigate this, but in my tests that often overshoots the results by a lot when merging from multiple different sources that have used different offsets.

One thing we can do make all relays use the same offset _for each query_ by making the offset be given by a hash of the `<subscription_id>`, for example (and then clients have to ensure they use the same subscription id for all relays they query). That is sad because it negates the possibility of having relays that only store an HLL value for each query and discards all events.

Another idea, maybe better, is to use a fixed offset like proposed here, but instead of using the event id, use the author pubkey. Although this makes it so some pubkeys will always have more weight than others when liking posts or following people, which is like creating a funny caste system within Nostr.

fiatjaf commented:
OK, I think I got the solution: make the offset dependent on the filter, not the subscription id. Something like

- if you are querying for reactions to event "abc123" the query will be `{"#e": ["abc123"], "kinds": [7]}`, then the offset would be given by "c1", for example.
- if you are querying for followers of pubkey "mno456" the query will be `{"#p": ["mno456"], "kinds": [3]}`, then the offset would be given by "o4", for example.
- and so on.

Just checking for the first "#e", #p", "#a" or "#q" tag (in this order) will cover 99% of use cases, but and if there is no unambiguous way to determine the offset, use a predefined hardcoded value.

v0l commented:
wow, nice!

vitorpamplona commented:
Very cool! But I do think this is highly gamefiable because Nostr is so open and the ID is not random at all. Even with the latest filter-based offsets, it's possible to run a PoW procedure on expected filters and significantly change the count. 

The procedure must use the same offset for all relays for the merging math to work. 

How about a sha256(id + subID) with the instruction that the subscription ID must be random and the same for all counting relays?

fiatjaf commented:
If you're using the subid there is no need to hash anything, you can just assume it will be random. If it's not that's the client's choice and their problem.

But I would still like something deterministic to allow for the storageless counter relay. I think the latest procedure plus using the event `pubkey` instead of event `id` is sufficient. People can't change their pubkey after the fact just to game one specific like count. And if the relay is going to accept events from completely new and random pubkeys then the counts from that relay are already borked and meaningless anyway.

mikedilger commented:
This is very cool.

I'm not sure the mod 24 part is doing what you intended.  The hex characters are lowercase so 'a'-'f' gives you 1-6 overlapping with what '1'-'6' gives you... which doesn't hurt but is less effective I think.

mikedilger commented:
Maybe something like this instead:

    byte offset in pubkey = filterbytes[16] AND 31   (16th byte of the filter data, mapped to a number from 0-31).

> People can't change their pubkey after the fact just to game one specific like count

True.  But if someone manages to get a pubkey with a hell of a lot of zeroes, many things that they react to will seem to be highly reacted to.

I think one more step fixes this. Instead of counting zeroes in the pubkey directly,  you count zeroes in the XOR of the pubkey and some hash that is generated from the filter query.

vitorpamplona commented:
> If you're using the subid there is no need to hash anything, you can just assume it will be random enough. If it's not that's the client's choice and their problem.

That's why I think the hash is needed. It would get rid of any PoW made for the event ID and whatever comes in as subID creates enough of a variance that the whole ID changes.

alexgleason commented:
@vitorpamplona If we take a hash they will just PoW the hash.

mikedilger commented:
Take a look at HLL-RB (HyperLogLog with random buckets). I think relays may be able to use a random number... even for the same query twice.

fiatjaf commented:
> True. But if someone manages to get a pubkey with a hell of a lot of zeroes, many things that they react to will seem to be highly reacted to.

We can get the offset from something like a xor between the pubkey and some part of the filter (I'd rather not take the filter JSON verbatim because that would have been preprocessed already in most cases, it would complicate implementations), then read the stuff from the event id based on that offset.

alexgleason commented:
I think it's impossible to create a deterministic identifier that can't be mined.

vitorpamplona commented:
> @vitorpamplona If we take a hash they will just PoW the hash.


Yep, that's why I suggested hash(event id + subid).

mikedilger commented:
> I think it's impossible to create a deterministic identifier that can't be mined.

We are talking about mining fresh pubkeys that would like your post.  Relays can use their spam prevention logic to reject such fresh pubkeys.

I think that adding randomness would not mess up the count (on a given relay) but it would make it impossible to combine counts between multiple relays (if you naively did, you would overshoot).

fiatjaf commented:
> I suppose they can do that a lot already by just generating thousands of events.

Exactly, so we are not adding any new weaknesses. Relying on reaction counts from relays that will accept anything is already a very bad idea.

> We are talking about mining fresh pubkeys that would like your post. Relays can use their spam prevention logic to reject such fresh pubkeys.

Agreed.

fiatjaf commented:
> I'm not sure the mod 24 part is doing what you intended. The hex characters are lowercase so 'a'-'f' gives you 1-6 overlapping with what '1'-'6' gives you... which doesn't hurt but is less effective I think.

I don't fully get this, but the mod 24 was in relation to the bytes, not to the hex characters. It was maximum 24 in order to leave room for 8 bytes at the end of the id/pubkey from which we would read from. I guess we should also skip the first 8 characters since they're so often used for generic PoW, that leaves us 16 possible values for the `offset`, which means we can read that from a single hex character.

mikedilger commented:
> > I'm not sure the mod 24 part is doing what you intended. The hex characters are lowercase so 'a'-'f' gives you 1-6 overlapping with what '1'-'6' gives you... which doesn't hurt but is less effective I think.
> 
> @mikedilger I don't fully get this, but the mod 24 was in relation to the bytes, not to the hex characters. It was maximum 24 in order to leave room for 8 bytes at the end of the id/pubkey from which we would read from. I guess we should also skip the first 8 characters since they're so often used for generic PoW, that leaves us 16 possible values for the `offset`, which means we can read that from a single hex character. I know this is all made in a very ad-hoc stupid way, but I like it because it's simple.

Ok then I think it is worded in a confusing way. It sounded like you were taking the hex characters as ascii, and doing a mod 24 on that.

mikedilger commented:
I couldn't get this to work when trying to copy the `go` code (probably my fault).  So I tried reading and implementing the 2007 paper and I got it to work.  However, these algorithms differ.

https://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf  (see the top of page 140).

The original 2007 algorithm uses `ρ(s)` as being the position of the first 1 bit, not the count of leading 0s, meaning the minimum value will be 1.  This can't just be corrected by a factor of 2 because it then changes the count of slots that are still 0 (every slot actually used will be at least 1).  (original LogLog set slots to -∞ to do this differentiation)

Relays could each use their own different estimation algorithms, but the 256 counts will have to be normalized to either be a count of zeroes, or the position of the first 1 (a count of zeroes +1) in order to be interoperable.

fiatjaf commented:
There are different implementations out there, I tried to do the simplest possible way that would work and wrote it on the NIP.

We will have to agree on something, we can't have each relay implementing their own way.



> The original 2007 algorithm uses `ρ(s)` as being the position of the first 1 bit, not the count of leading 0s, meaning the minimum value will be 1.

This is what my implementation does too, it counts zeroes and adds 1.

mikedilger commented on 45.md:
```diff
@@ -29,29 +29,65 @@ In case a relay uses probabilistic counts, it MAY indicate it in the response wi
+
+This section describes the steps a relay should take in order to return HLL values to clients.
+
+1. Upon receiving a filter, if it has a single `#e`, `#p`, `#a` or `#q` item, read its 32th ascii character as a byte and take its modulo over 24 to obtain an `offset` -- in the unlikely case that the filter doesn't meet these conditions, set `offset` to the number 16;
```
1. Upon receiving a filter, if it has a `#e`, `#p`, `#a` or `#q` item with a single value (checked in that order in case there are multiple of these), then covert that value into 32 bytes (for the 'a' tag use the pubkey part only) and take the 16th byte modulo 24 as an 'offset'.  In the unlikely case that the filter doesn't meet these conditions, set `offset` to the number 16;

mikedilger replied:
eh, for the 'a' tag I guess it doesn't matter, the kind number isn't too long so chars 32 and 33 are still inside the hex part.

fiatjaf replied:
So instead of fixing this I doubled down on my error and took the opportunity to exclude the first 8 bytes of all pubkeys. If you say this is stupid I'll do as you suggested.

mikedilger replied:
Fine by me. The offset isn't that important.  The original algorithm uses offset=0 on hash outputs.

mikedilger commented on 45.md:
```diff
@@ -29,29 +29,65 @@ In case a relay uses probabilistic counts, it MAY indicate it in the response wi
+2. Initialize 256 registers to 0 for the HLL value;
+3. For all the events that are to be counted according to the filter, do this:
+    1. Read byte at position `offset` of the event `pubkey`, its value will be the register index `ri`;
+    2. Count the number of leading zero bits starting at position `offset+1` of the event `pubkey`;
```
2. Count the number of leading zero bits starting at position 'offset+1' of the event 'pubkey' and add 1;

fiatjaf replied:
Fixed.

mikedilger reviewed (changes requested):
This is just a restatement of my prior comments as edits

mikedilger commented:
Because NIP-45 COUNT counts the number of events, but hll counts the number of distinct pubkeys, I think perhaps the key should be "hll_pubkeys" instead of just "hll" to at least give a hint that it isn't counting the events that match the filter, and to leave room for an "hll" that does.

fiatjaf commented:
I think in all cases so far what we actually want is distinct pubkeys, so we can be pragmatic about it.

If we ever make a different thing for the the number of distinct ids we can come up with a different name, like `hlli`, I don't know, these keys are all hardcoded anyway.

I would prefer a JSON array over an object with keys because then we wouldn't have these discussions.

mikedilger commented:
> What is the expected general relay behavior?

The straightforward solution is:
1) Find all the matching events
2) Process them through the hll algorithm
3) Send a count of them, along with the hll output (512 characters representing 256 small integers).
4) Don't cache anything.

Caching is hard because
1) The 32-byte value of the filter is not the only part of the filter.  You can't presume the next filter that passes the hll appropriateness test also has the same other filter fields.  And you can't easily hash the entire filter (unless your relay has  a deterministic way to represent filters).
2) Every time a new event comes in... for each of the cached hlls.... you have to test the event against that filter (you need to remember the filters) and then update the hll data (itself being the easy and quick part).

IMHO counting should already be very fast since it uses indexes and isn't copying data or allocating memory, especially if the data being counted is mmapped.

fiatjaf commented on 45.md:
```diff
@@ -29,29 +29,65 @@ In case a relay uses probabilistic counts, it MAY indicate it in the response wi
+
+### `hll` encoding
+
+The value `hll` value must be the concatenation of the 256 registers, each being a uint8 value (i.e. a byte). Therefore `hll` will be a 512-character hex string.
```
No.

fiatjaf commented on 45.md:
```diff
@@ -29,29 +29,65 @@ In case a relay uses probabilistic counts, it MAY indicate it in the response wi
+
+### Algorithm
+
+This section describes the steps a relay should take in order to return HLL values to clients.
+
+1. Upon receiving a filter, if it has a single `#e`, `#p`, `#a` or `#q` item, read its 32th ascii character as a nibble (a half-byte, a number between 0 and 16) and add `8` to it to obtain an `offset` -- in the unlikely case that the filter doesn't meet these conditions, set `offset` to the number `16`;
+2. Initialize 256 registers to `0` for the HLL value;
+3. For all the events that are to be counted according to the filter, do this:
+    1. Read byte at position `offset` of the event `pubkey`, its value will be the register index `ri`;
+    2. Count the number of leading zero bits starting at position `offset+1` of the event `pubkey` and add `1`;
+    3. Compare that with the value stored at register `ri`, if the new number is bigger, store it.
```
"Why all this complexity?"

"A new extension can be added..."

mikedilger replied:
The way I would do it if the user provides a some random data in the new or modified COUNT command is to hash (or just xor) that data with the resultant event pubkey and start from bit 0 (no offset).  BUT this eliminates the possibility of the relay caching the result.

fiatjaf commented:
> Caching is hard

Caching is harder, but I think it can be done specially if you want to store counts for a huge number of events using almost no disk space.

Storing millions of reactions is incredibly wasteful.

fiatjaf commented:
The usage of registers has nothing to do with IP addresses. Any user will be able to skyrocket the number if they can generate infinite number of pubkeys and issue likes/follows/etc with all these pubkeys. If you can generate 3000 pubkeys you can make the follow count go up by 3000. That is true today, will be true with this change, will always be true. We have no way around it except 

1. downloading a bazillion number of events from public relays and filtering locally using WoT or whatever;
2. relying on relays that are known to do a good job in filtering stuff -- paid relays and WoT relays from reputable humans, for example -- then merging their HLL values locally.

If HLL values are cached on the relay side or counted at request time is irrelevant for the client and the results will be exactly the same.

mikedilger commented:
In order to game this system, you have to produce close to 256 pubkeys each with effective PoW of about 24 (to reach 8 million likes, for example) and these newly mined pubkeys would have to be accepted by the target relay and not seen as spam-pubkeys.  IMHO this is not too hard to game.  It just takes a little effort and ingenuity.

But how severe is that?  So what if it looks like millions of people upvoted you. Reactions are stupid anyways.  It's not like they are forging your events or reading your DMs. And if clients look at the first results from the COUNT commands they can quite easily detect the HLL abuse.

Still... as this is quite easy to repair as I've said by having some kind of randomness (perhaps supplied by the client with the COUNT is best to make it consistent across relays without being known to the attacker gaming things).  But again the downside is that relays can't cache things anymore (or they can, but probably won't get cache hits anymore) and they have to keep the original reaction events. But at least clients don't need those events anymore.

I don't have a strong opinion either way, but I lean towards having the client-supplied random thing, because I have a penchant for reliability and trustability moreso than efficiency.  I think I've said all I can say about this so I'll stop commenting on it until or unless we start debating particulars after this choice is made.

vitorpamplona commented:
Why can't the relay count and ALSO run HLL to make sure values match? 

Relays MUST guarantee that the HLL is under x% of the actual number and not return an HLL if it goes over.

This seems like a better protection than these shenanigans to try to block gamification.

vitorpamplona commented:
Yep, but the client cannot fix the HLL if it's wrong. The relay can automatically find which event IDs are creating the distortion and fix it (ignore them) before sending it back.

vitorpamplona commented:
> Combining the event ID or the pubkey of the author with a client nonce means it is not possible to intentionally bias pubkeys or event IDs

But I don't see any real nonce in this spec. Or are we back to using the subscription ID as a random nonce?

vitorpamplona commented:
I am in favor of the client-based nonce. It's the only way to keep the counter usable.

fiatjaf commented:
Why not 3 ways? 4 ways? 17 ways? Why have a standard at all? Let people do what they want!

fiatjaf commented:
OK, I tried to build the cache thing and it's too complicated. Nostr query system is too flexible. It will never work to support open-ended queries, it will end up using more space than storing the actual events.

It's tempting to just declare caching a failed idea and standardize hyperloglog without that possibility, but it would be so good to not have to store millions of reactions and that stuff!

Now I wonder if we should standardize just some of the "basic" queries to support cached hyperloglog, like number of reactions, number of quotes, number of comments, number of followers -- so these and only these could be cached, and for all the rest (and for all the subqueries of these or other complexities) relays have to either count at runtime or not fulfill at all?

What are the "basic" count queries, besides the ones I listed above?

Or am I crazy to insist in caching?

vitorpamplona commented:
> am I crazy to insist in caching?

This seems very similar to negentropy. Caching is just not viable.

mikedilger commented:
I think the idea of trying to avoid storing of many low-value events on relays is laudable.  I think we should spend more time trying to figure that out.

If we still need to return a normal COUNT though, we will at least need to store the ID prefixes to avoid counting duplicates.  And we will need a normal COUNT result in order to detect gaming of the system.

In order for relays to keep these HLL counts without keeping the events, they need to start counting them for well-specified filters before anybody even asks for one.  So I agree that we need to specify well-known filters representing things to count (e.g. number of reactions, number of followers).  And in that case we cannot hash the event value (pubkey or id) with a user-supplied nonce.  And since all parties need to use the same method of generating HLLs for them to combine properly, the system will be well-known and thus gameable.  But outliers can be detected and rejected, preventing most gaming.

I'm almost in favor of doing this both ways... but no more.  Not 17 ways.  I could be persuaded to do it 2.718 ways.

mikedilger commented:
Or instead of keeping all the ID prefixes, you keep a bloom filter and when a "definitely not in set" happens you add 1 but when a "possibly in set" happens you add the false positive rate instead.

fiatjaf commented:
> dedicated indexes and encoding schemes can be made for high-volume optimizable events

HLL is exactly such a thing already and defining what goes into this dedicated scheme and what is ignored is the question I posed above: if you want a limited functionality for specific use cases then HLL caching can be very good, otherwise nothing is possible.

fiatjaf commented:
> You have not solved the problem that this is open to manipulation

I am pretty sure I did: do not trust counts from relays that accept events from random keys. It's literally the same problem with or without HLL.

fiatjaf commented:
> Should a new user's reaction count? Are they a "random key"? How do you distinguish new users from attempts at manipulation if they are spread over multiple IPs?

If our only tool for fighting spam is IP address detection then I'm sorry for us. I don't even consider IP as a way to identify spam at all.

Much better solutions are paid relays, WoT relays, captcha relays, relays that use any other algorithm for checking what is a "valid" key like activity traacking or whatever, or any other form of whitelisted relays.

fiatjaf commented:
OK, I changed this to specify only a hardcoded set of filters eligible for HyperLogLog.

I've implemented it on khatru and eventstore and deployed it to https://pyramid.fiatjaf.com/.

Pyramid is using a disk cache for `kind:7` reactions (it doesn't store the events) and a just-in-time calculation for `kind:3` (it does store the events), so this is nice and flexible.

(This has just help me find an unrelated bug on khatru, by the way.)

mikedilger commented:
> no real solution here to the attack still

I don't know what you mean by "real", but Fiatjaf proposed many means of avoiding rogue gamed pubkeys in this comment: https://github.com/nostr-protocol/nips/pull/1561#issuecomment-2494724003

Plus I suggested a way to avoid it by having clients send a nonce in with the COUNT (they would use the same nonce on all the relays so that the HLLs would properly merge, but it wouldn't be known ahead of time so it can't be gamed).

> This method can still be implemented with a "secure" HLL with a pre-determined key.

No, if the key is known (pre-determined) then I can mine a pubkey designed to create an artificially high HLL count for it's bucket against that pre-determined known key.

bezysoftware commented:
Where would the nonce go in the request? The way it's structured now (trailing list of filter objects) doesn't really allow a place to put it. Should we take subscription ID as nonce?

vitorpamplona commented:
I don't think the PR, as it stands right now, is that useful. So, nack for me. But I don't oppose merging it if this is useful for somebody. 

> I suggested a way to avoid it by having clients send a nonce in with the COUNT

IMO that's the only realistic way to use HLL results.

fiatjaf commented:
> The way it's structured now (trailing list of filter objects) doesn't really allow a place to put it. Should we take subscription ID as nonce?

The subscription id is autogenerated as an arbitrary by most libraries today, so it would be nasty to put it there and then require clients to use the same subscription id for every relay, so it's better to put it outside somewhere.

I think it would be ok to make a BREAKING CHANGE and require COUNT requests to have a single filter. It's hard to believe anyone is using COUNT with more than one filter today, it's even hard to believe anyone is using COUNT at all.

mikedilger commented:
> Where would the nonce go in the request? The way it's structured now (trailing list of filter objects) doesn't really allow a place to put it. Should we take subscription ID as nonce?

I new form of COUNT would have to be defined.

mikedilger commented:
Implemented in chorus https://github.com/mikedilger/chorus/commit/964d68de5768abeeaccad06f425ec9c3f199faf7

fiatjaf commented:
Implemented it on nak: https://github.com/fiatjaf/nak/commit/7f608588a2616c52f93aab1becfb6aca954e6c35

Here are some test runs using HLL output from _pyramid.fiatjaf.com_ and _chorus.mikedilger.com:444_:

```bash
~> nak count -k 3 -p a1fc5dfd7ffcf563c89155b466751b580d115e136e2f8c90e8913385bbedb1cf pyramid.fiatjaf.com chorus.mikedilger.com:444
wss://pyramid.fiatjaf.com: 13 📋
wss://chorus.mikedilger.com:444: 13 📋
📋 HyperLogLog sum: 15
~> nak req -k 3 -p a1fc5dfd7ffcf563c89155b466751b580d115e136e2f8c90e8913385bbedb1cf pyramid.fiatjaf.com chorus.mikedilger.com:444 | jq '.pubkey' | sort | uniq | wc -l
16
```

```bash
~> nak count -k 3 -p 3bf0c63fcb93463407af97a5e5ee64fa883d107ef9e558472c4eb9aaaefa459d py
ramid.fiatjaf.com chorus.mikedilger.com:444
wss://pyramid.fiatjaf.com: 110 📋
wss://chorus.mikedilger.com:444: 144 📋
📋 HyperLogLog sum: 152
~> nak req -k 3 -p 3bf0c63fcb93463407af97a5e5ee64fa883d107ef9e558472c4eb9aaaefa459d pyra
mid.fiatjaf.com chorus.mikedilger.com:444 | jq '.pubkey' | sort | uniq | wc -l
166
```

```bash
~> nak count -k 3 -p 32e1827635450ebb3c5a7d12c1f8e7b2b514439ac10a67eef3d9fd9c5c68e245 pyramid.fiatjaf.com chorus.mikedilger.com:444
connecting to pyramid.fiatjaf.com... ok.
connecting to chorus.mikedilger.com:444... ok.
wss://pyramid.fiatjaf.com: 140 📋
wss://chorus.mikedilger.com:444: 141 📋
📋 HyperLogLog sum: 158
~> nak req -k 3 -p 32e1827635450ebb3c5a7d12c1f8e7b2b514439ac10a67eef3d9fd9c5c68e245 pyramid.fiatjaf.com chorus.mikedilger.com:444 | jq '.pubkey' | sort | uniq | wc -l
156
```
